### ISSUE TITLE
Trouble closing Dask comms

### ISSUE THREAD

**Issue by mrocklin** [rapidsai/ucx-py#100]
This is possibly a problem on the Dask side, but any help interpretting the error message would be welcome.  cc @Akshay-Venkatesh @rjzamora 

```
from dask.distributed import LocalCluster
import asyncio

async def f():
    async with LocalCluster(protocol='ucx', interface='ib0', asynchronous=True,
            dashboard_address=None, data=dict, n_workers=1) as cluster:
        pass


if __name__ == '__main__':
    asyncio.get_event_loop().run_until_complete(f())
```

```
[1556829182.493231] [dgx15:4386 :0]     ucp_worker.c:495  UCX  ERROR error 'Endpoint timeout' will not be handled for ep 0x7f7ef7627048 - rc_mlx5/mlx5_3:1
[dgx15:4386 :0:4386] ib_mlx5_log.c:116  FATAL: Error on mlx5_3:1 QP 0x11987 wqe[003]: Transport retry count exceeded (synd 0x15 vend 0x81 hw_synd 0/0) opcode SEND
==== backtrace ====
    0  /home/nfs/mrocklin/miniconda/envs/ucx-dev/lib/libucs.so.0(ucs_log_default_handler+0x267) [0x7f7f239304d7]
    1  /home/nfs/mrocklin/miniconda/envs/ucx-dev/lib/libucs.so.0(ucs_log_dispatch+0xe1) [0x7f7f239306f1]
    2  /home/nfs/mrocklin/miniconda/envs/ucx-dev/lib/libuct.so.0(uct_ib_mlx5_completion_with_err+0x1f6) [0x7f7f23d985e6]
    3  /home/nfs/mrocklin/miniconda/envs/ucx-dev/lib/libuct.so.0(uct_rc_mlx5_iface_progress+0x5ce) [0x7f7f23dac91e]
    4  /home/nfs/mrocklin/miniconda/envs/ucx-dev/lib/libucp.so.0(ucp_worker_progress+0x4a) [0x7f7f24038e8a]
    5  /home/nfs/mrocklin/ucx-py/ucp/_libs/ucp_py.cpython-37m-x86_64-linux-gnu.so(+0xad43c) [0x7f7f2433d43c]
    6  /home/nfs/mrocklin/ucx-py/ucp/_libs/ucp_py.cpython-37m-x86_64-linux-gnu.so(ucp_py_put_ep+0x48) [0x7f7f2433dc88]
    7  /home/nfs/mrocklin/ucx-py/ucp/_libs/ucp_py.cpython-37m-x86_64-linux-gnu.so(+0x46d6a) [0x7f7f242d6d6a]
    8  /home/nfs/mrocklin/ucx-py/ucp/_libs/ucp_py.cpython-37m-x86_64-linux-gnu.so(+0x261b5) [0x7f7f242b61b5]
    9  /home/nfs/mrocklin/ucx-py/ucp/_libs/ucp_py.cpython-37m-x86_64-linux-gnu.so(+0x289eb) [0x7f7f242b89eb]
   10  python(_PyMethodDef_RawFastCallKeywords+0x8c) [0x563eb5e2c50c]
   11  python(_PyCFunction_FastCallKeywords+0x21) [0x563eb5e2c801]
   12  python(_PyEval_EvalFrameDefault+0x4e8c) [0x563eb5e882bc]
   13  python(_PyFunction_FastCallKeywords+0xfb) [0x563eb5e2b79b]
   14  python(_PyEval_EvalFrameDefault+0x6a0) [0x563eb5e83ad0]
   15  python(_PyGen_Send+0x2a2) [0x563eb5e25672]
   16  /home/nfs/mrocklin/miniconda/envs/ucx-dev/lib/python3.7/lib-dynload/_asyncio.cpython-37m-x86_64-linux-gnu.so(+0xc18e) [0x7f7f24c3a18e]
   17  python(_PyObject_FastCallKeywords+0x49b) [0x563eb5e2cd2b]
   18  python(+0x20abc3) [0x563eb5ebdbc3]
   19  python(_PyMethodDef_RawFastCallDict+0x199) [0x563eb5def8b9]
   20  python(_PyCFunction_FastCallDict+0x21) [0x563eb5defae1]
   21  python(_PyEval_EvalFrameDefault+0x5d21) [0x563eb5e89151]
   22  python(_PyFunction_FastCallKeywords+0xfb) [0x563eb5e2b79b]
   23  python(_PyEval_EvalFrameDefault+0x6a0) [0x563eb5e83ad0]
   24  python(_PyFunction_FastCallKeywords+0xfb) [0x563eb5e2b79b]
   25  python(_PyEval_EvalFrameDefault+0x6a0) [0x563eb5e83ad0]
   26  python(_PyFunction_FastCallKeywords+0xfb) [0x563eb5e2b79b]
   27  python(_PyEval_EvalFrameDefault+0x6a0) [0x563eb5e83ad0]
   28  python(_PyFunction_FastCallKeywords+0xfb) [0x563eb5e2b79b]
   29  python(_PyEval_EvalFrameDefault+0x6a0) [0x563eb5e83ad0]
   30  python(_PyEval_EvalCodeWithName+0x2f9) [0x563eb5dc94f9]
   31  python(PyEval_EvalCodeEx+0x44) [0x563eb5dca3c4]
   32  python(PyEval_EvalCode+0x1c) [0x563eb5dca3ec]
   33  python(+0x22f874) [0x563eb5ee2874]
   34  python(PyRun_FileExFlags+0xa1) [0x563eb5eecb81]
   35  python(PyRun_SimpleFileExFlags+0x1c3) [0x563eb5eecd73]
   36  python(+0x23ae5f) [0x563eb5eede5f]
   37  python(_Py_UnixMain+0x3c) [0x563eb5eedf7c]
   38  /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0) [0x7f7f26f49830]
   39  python(+0x1e0122) [0x563eb5e93122]
===================
```

**Comment by Akshay-Venkatesh** [comment]
My guess is that this problem appears because there's a send/recv outstanding that wasn't satisfied. Is it possible that dask posted a recv for a possible message in the future but the job finished?

**Comment by Akshay-Venkatesh** [comment]
If I'm not wrong, there are at the very least heartbeat messages exchanged which can theoretically go unsatisfied at some point.

**Comment by mrocklin** [comment]
This test fails

```
(ucx-dev) mrocklin@dgx15:~/ucx-py$ py.test tests/test_multiple_listeners.py  -s
============================================ test session starts =============================================
platform linux -- Python 3.7.3, pytest-4.4.1, py-1.8.0, pluggy-0.9.0
rootdir: /home/nfs/mrocklin/ucx-py
plugins: asyncio-0.10.0
collecting ... /home/nfs/mrocklin/miniconda/envs/ucx-dev/bin/python: symbol lookup error: /home/nfs/mrocklin/miniconda/envs/ucx-gpu/lib/ucx/libuct_ib.so.0: undefined symbol: cuCtxGetDevice
```

```python
import ucp
import asyncio

ucp.init()


async def hello(ep, listener):
    await asyncio.sleep(0)
    ep.close()
    ucp.stop_listener(listener)


def test_multiple_listeners():
    lf1 = ucp.start_listener(hello, is_coroutine=True)
    lf2 = ucp.start_listener(hello, is_coroutine=True)
    assert lf1.port > 0
    assert lf2.port > 0
    assert lf1.port != lf2.port

    lf1.coroutine.close()
    lf2.coroutine.close()
    ucp.stop_listener(lf1)
    ucp.stop_listener(lf2)
```

More generally, can you verify that tests pass for you?

**Comment by mrocklin** [comment]
Ah, whoops, that last comment is in the wrong issue. Moving.

**Comment by mrocklin** [comment]
> Is it possible that dask posted a recv for a possible message in the future but the job finished?

Yes, this is possible.  Dask has an open recv on the server side.  It recvs and then waits until something sends into it.  Ideally the sending side sends a close signal, but I wouldn't be surprised if it didn't.  Would it be possible to fail more gracefully here?  I can look for cases where comms may not clean up well in the meantime.

**Comment by mrocklin** [comment]
Just to be clear, is it feasible for UCX to fail more gracefully here?  It's possible that I can avoid this in some cases but in principle there are going to be missed communications.  It would be good if we could avoid halting the process.

**Comment by mrocklin** [comment]
For information here, we seem to fail on this line:

            ucp.destroy_ep(self._ep)

https://github.com/dask/distributed/pull/2591/files#diff-ca5c24c584111a291f763107e4065539R207

I would be very happy to tell the endpoint "no one is coming for you, it's ok to give up" if that was an option.

**Comment by mrocklin** [comment]
I've cleaned things up a bit in Dask.  Those changes are now both in dask/master and quasiben/tom-ucx.  Things usually work ok if workers are never killed.  However, if a worker is killed unexpectedly (which we should expect) then anyone that was waiting on a connection from that worker can also die unhappily.  If we care about resilience then this would be a good thing to track long term.

So from my perspective short term we're ok, but long term this is an issue.

**Comment by Akshay-Venkatesh** [comment]
@mrocklin It seems like the approach we may want for resilience is track outstanding communication requests and [cancel](https://github.com/openucx/ucx/blob/master/src/ucp/api/ucp.h#L2867) outstanding requests if they're not satisfied by the time the a) endpoint is closed or b) ucp context is destroyed. We'll need some tracking logic to associate requests with endpoints for case a.

**Comment by Akshay-Venkatesh** [comment]
I'll work on a PR for this if the above approach suffices. Lmk if otherwise.

**Comment by mrocklin** [comment]
Update here on my end is that Dask is better about cleaning things up so this should be an issue less frequently.  Has anything changed on the UCX side @Akshay-Venkatesh ?  If so, can we close?

**Comment by Akshay-Venkatesh** [comment]
@mrocklin I had been seeing intermittent failures when running `tests/test_send_recv_obj.py` about:
1. mpool requests not released 
2. rcache refcount not being 0

This [commit](https://github.com/rapidsai/ucx-py/pull/128/commits/7e390ef5ad7490e2e9cfd914c8f81fdb301dd049) fixes those issues. I had to ensure that I wasn't calling `ucp.fin()` before listener and client had a chance to release resources. Calling await on `listener.coroutine` task proved important. Can you let me know if this is relevant to dask? Thanks

**Comment by quasiben** [comment]
I am still running into this issue -- on latest distributed and ucx-py

**Comment by quasiben** [comment]
I believe i am currently seeing a few lingering tasks and destroying the endpoint causes the process to abort

**Comment by jakirkham** [comment]
Also saw this error recently when trying to run the "My Benchmark" code in [this blogpost]( https://blog.dask.org/2019/06/09/ucx-dgx ). Though can't claim the latest UCX or UCX-Py versions. However do have the latest Dask and Distributed versions.

**Comment by jakirkham** [comment]
@quasiben, are you still seeing this or were you able to workaround/solve this issue somehow?

**Comment by quasiben** [comment]
Still seeing this but I have deeply investigated what is happening

**Comment by pentschev** [comment]
> Still seeing this but I have deeply investigated what is happening

I think you meant you have _not_?

**Comment by quasiben** [comment]
yes...

**Comment by quasiben** [comment]
removing now that #180 is merged in -- if there are still issues closing comms we should open a new issue
